enhancements/aws-pod-identity.md

```diff
@@ -0,0 +1,442 @@
+---
+title: aws-pod-identity 
+authors:
+  - "@sjenning"
+reviewers:
+  - "@derekwaynecarr"
+  - "@smarterclayton"
+approvers:
+  - "@derekwaynecarr"
+creation-date: 2020-03-26
+last-updated: 2020-03-26
+status: provisional
+see-also:
+  - ""  
+replaces:
+  - ""
+superseded-by:
+  - ""
+---
+
+# Support Native (IAM-based) Pod Identity on AWS
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+## Summary
+
+When deployed on AWS, Kubernetes can use pod identity managed by IAM.  AWS has a
+[webhook](https://github.com/aws/amazon-eks-pod-identity-webhook/) they maintain
+for EKS that does enables this ability.  This enhancement integrates the same
+functionality into OpenShift.
+
+The `cloud-credential-operator` will be responsible for managing this new
+ability as it serves as an alternative credential _injection_ mechanism versus
+the existing mechanism of injecting an IAM user acccess ID and key into a
+secret.  The existing role/user _creation_ via a CredentialRequest CR is common
+between the methods.
+
+The new injection mechanism works by creating a `ServiceAccount` annotated with
+a IAM role the pod should assume. When a pod runs with the annoated
+`ServiceAccount`, the webhook injects a projected volume into the pod that
+contanis the bound token that can be used to assume the IAM role.  The AWS SDK
+searches for this token automatically.
+
+## Motivation
+
+This allows users to more cleanly express and manage pod identities in IAM when
+running on AWS.  The use of bound tokens to assume roles (vs user id/key) is
+preferred as these tokens can expire and be rotated.
+
+### Goals
+
+1. A pod can assume an IAM role created with a `CredentialRequest` CR using a
```

Comment:
So this means that a pod can have only ONE identity attached right?

cloud-credential-operator currently supports dual identity.
- an admin entity that is used to create new roles/users etc.
- an reader entity that is only allowed to read roles/users to make sure things stay as-is


a lot of the operators today require create/update/delete permissions do that they can create dynamic objects. But some of the operators can be modified to have dual identities.
- ADMIN_WRITE identity that can be used to create the objects.
- RESTRICTED_WRITE identity that is restricted to only the objects it created.
that could allow users to remove/disable the ADMIN_WRITE permissions as day-2 operation.. making the security stance even stronger.

if using the POD identity based on webhook and environment disallows us to ever move in that direction, I think we should maybe not use it and stick to secrets as multiple secrets can be mounted compared to single AWS env variable.

Reply:
yes, this would currently not be suitable for use by the CCO itself.  since the pod can only run with one serviceaccount, this mechanism is one-to-one IAM role to pod.

Reply:
would/could this be solved by having the operator have the ADMIN_WRITE and the operand have RESTRICTED_WRITE (assuming a clean separation between operator and operand)? 

i suppose this might also imply that the operator is also creating a new CredentialsRequest so that the perms for the operand only grant perms to the actually created objects...

Reply:
i think we should punt on the multi-cred use case for now since serviceaccount-to-pod is 1:1 vs secret-to-pod is N:1